### PR TITLE
S5: Simplified State Space Layers for Sequence Modeling

### PR DESCRIPTION
**Pull request by emailweixu**
https://openreview.net/forum?id=Ai8Hw3AXqks

Tested on sMNIST got similar result as running the original S5 code.

Results from this PR
Train Loss: 0.00022 -- Val Loss: 0.03176 --Test Loss: 0.02213 -- Val Accuracy: 0.9955 Test Accuracy: 0.9954

Results from the original code
Train Loss: 0.00119 -- Val Loss: 0.02077 --Test Loss: 0.01888 -- Val Accuracy: 0.9950 Test Accuracy: 0.9954

### CODE REVIEW

File: alf/networks/s5/utils.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+    """
+    # Make -HiPPO
+    hippo = make_HiPPO(N)
+
+    # Add in a rank 1 term. Makes it Normal.
+    P = np.sqrt(np.arange(N) + 0.5)
+
+    # HiPPO also specifies the B matrix
+    B = np.sqrt(2 * np.arange(N) + 1.0)
```
Comment (breakds) [inline alf/networks/s5/utils.py:51]: Seems `B = np.sqrt(2) * P`

File: alf/networks/s5/s5.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+
+    Note: this implementation corresponds to s5.ssm.S5SSM with the following settings:
+
+        C_init='lecun_normal'
+        bidirectional=False
+        conj_sym=True
+        clip_eigs=True
+        discretization='zoh'
+
+    Args:
```
Comment (breakds) [inline alf/networks/s5/s5.py:140]: missing `num_blocks` in `Args:`.

File: alf/networks/s5/s5.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+                 dt_max: float = 0.1,
+                 step_rescale: float = 1.0,
+                 name="S5"):
+        super().__init__(
+            input_tensor_spec=alf.TensorSpec((data_dim, )),
+            state_spec=alf.TensorSpec((state_dim, )),
+            name=name)
+        assert state_dim % 2 == 0
+        assert state_dim // 2 % num_blocks == 0
+        cstate_dim = state_dim // 2
```
Comment (breakds) [inline alf/networks/s5/s5.py:167]: Would be helpful to explain that `c` here means complex.

File: alf/networks/s5/s5.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+    Corresponds to s5.layers.SequenceLayer
+
+    Note: for sequence input, the input is assumed to be [T, B, C]
+
+    """
+
+    def __init__(self,
+                 ssm_ctor: Callable[..., nn.Module],
+                 dropout: float,
+                 activation: str = "half_glu2",
```
Comment (breakds) [inline alf/networks/s5/s5.py:337]: Need to explain the activation here, e.g. what are the possible options and their effects.

File: alf/networks/s5/s5.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+    """
+    Args:
+        x: real tensor with shape [..., N]
+    Returns:
+        complex tensor with shape [..., N // 2]
+    """
+    return torch.view_as_complex(x.reshape(*x.shape[:-1], x.shape[-1] // 2, 2))
+
+
+def _view_as_real(x):
```
Comment (breakds) [inline alf/networks/s5/s5.py:58]: Why do we want to view complex numbers as real, if all the operators can handle complex numbers (or maybe they cannot)?

File: alf/networks/s5/utils.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    # Store the final gradients for s and Lambda
+    tl.store(grad_s_ptr + col_offsets * 2, grad_s_real, mask=mask)
+    tl.store(grad_s_ptr + col_offsets * 2 + 1, -grad_s_imag, mask=mask)
+    tl.store(grad_lambda_ptr + col_offsets * 2, grad_lambda_real, mask=mask)
+    tl.store(
+        grad_lambda_ptr + col_offsets * 2 + 1, -grad_lambda_imag, mask=mask)
+
+
+class _ssm_forward(torch.autograd.Function):
+    BLOCK_SIZE = 128
```
Comment (breakds) [inline alf/networks/s5/utils.py]: How is the block size picked?

File: alf/networks/s5/utils.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+        grid = lambda meta: (triton.cdiv(n, meta['BLOCK_SIZE']), )
+
+        if Lambda.dtype == torch.complex64:
+            diag_ssm_forward_kernel_complex[grid](torch.view_as_real(s),
+                                                  torch.view_as_real(x),
+                                                  torch.view_as_real(y),
+                                                  torch.view_as_real(Lambda),
+                                                  length, batch_size, dim,
+                                                  _ssm_forward.BLOCK_SIZE)
+        else:
```
Comment (breakds) [inline alf/networks/s5/utils.py]: Slightly safer to have `elif Lambda.dtype.is_floating_point`, to avoid sending a `torch.complex128` here. And then with another `else` to report the error.

File: alf/networks/s5/utils.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+
+    P = V.conj().T @ P
+    B_orig = B
+    B = V.conj().T @ B
+    return Lambda_real + 1j * Lambda_imag, P, B, V, B_orig
+
+
+@triton.jit
+def diag_ssm_forward_kernel(s_ptr, x_ptr, lambda_ptr, y_ptr, length,
+                            batch_size, dim, BLOCK_SIZE: tl.constexpr):
```
Comment (breakds) [inline alf/networks/s5/utils.py]: Suggesting adding some documentation here about the "shape" of the `_ptr`s, using `length`, `batch` and `dim`

File: alf/networks/s5/s5.py [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+    Note: this implementation corresponds to s5.ssm.S5SSM with the following settings:
+
+        C_init='lecun_normal'
+        bidirectional=False
+        conj_sym=True
+        clip_eigs=True
+        discretization='zoh'
+
+    Args:
+        data_dim: dimension of input (H)
```
Comment (breakds) [inline alf/networks/s5/s5.py]: "dimension of input and output  (H)" to provide more information?

File: alf/networks/s5/s5.py [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+        ds/dt = As + Bu
+        A = V \Lambda V^{-1}
+        ds/dt = V \Lambda V^{-1}s + Bu
+        dV^{-1}s/dt = \Lambda V^{-1}s + V^{-1}Bu
+        y = C x + D u
+        y = CV V^{-1}s + Du
+
+    where :math:`s` is the state, :math:`u` is the input, :math:`y` is the output,
+    :math:`A` is the state transition matrix, :math:`B` is the input matrix,
+    :math:`C` is the output matrix, :math:`D` is the feedforward matrix,
```
Comment (breakds) [inline alf/networks/s5/s5.py:119]: Maybe mention `D` her is also assumed to be diagonal?

File: alf/networks/s5/utils.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+
+def diag_ssm_forward(s, x, Lambda):
+    r"""Diagonal SSM forward pass
+
+    Calculate :math:`y_t = Lambda * y_{t-1} + x_t` for t > 0
+    and :math:`y_0 = Lambda * s + x_0`
+
+    Args:
+        s (torch.Tensor): shape is [batch_size, state_dim]
+        x (torch.Tensor): shape is [length, batch_size, data_dim]
```
Comment (breakds) [inline alf/networks/s5/utils.py]: Here `x` comes from Bu, whose last dimension seems to be `state_dim`?

File: alf/networks/s5/utils.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+    @staticmethod
+    def backward(ctx, grad_y):
+        s, y, Lambda = ctx.saved_tensors
+        length, batch_size, dim = y.shape
+        grad_y = grad_y.contiguous()
+        n = batch_size * dim
+        grad_s = torch.empty_like(s)
+        grad_x = torch.empty_like(grad_y)
+        grad_lambda = torch.empty_like(s)
```
Comment (breakds) [inline alf/networks/s5/utils.py]: Nice trick. I am not sure whether this is very commonly used - but I was assuming `grad_lambda`'s shape is `[dim]` instead of `[B, dim]` until I saw the `grad_lambda.sum` below and fully understood the trick.

Suggest documenting the intention of making this temporary `grad_lambda` the same shape as `s` here.

File: alf/networks/s5/utils.py [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+
+
+@triton.jit
+def diag_ssm_backward_kernel_complex(
+        s_ptr, lambda_ptr, y_ptr, grad_s_ptr, grad_x_ptr, grad_lambda_ptr,
+        grad_y_ptr, length, batch_size, dim, BLOCK_SIZE: tl.constexpr):
+
+    # autograd for complex numbers calculates \partial f / \partial z^*
+    # so we need to take conjugate during the calculation.
+    # https://pytorch.org/docs/stable/notes/autograd.html#autograd-for-complex-numbers
```
Comment (breakds) [inline alf/networks/s5/utils.py]: Suggesting moving the comments here to right above line 225 for better readability-locality.

File: alf/networks/s5/utils.py [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+    Lambda = tl.load(lambda_ptr + col_offsets % dim, mask=mask, other=0)
+    for t in range(length):
+        offsets = t * batch_size * dim + col_offsets
+        x = tl.load(x_ptr + offsets, mask=mask, other=0)
+        s = s * Lambda + x
+        tl.store(y_ptr + offsets, s, mask=mask)
+
+
+@triton.jit
+def diag_ssm_backward_kernel(s_ptr, lambda_ptr, y_ptr, grad_s_ptr, grad_x_ptr,
```
Comment (breakds) [inline alf/networks/s5/utils.py]: Mathematically working out the backward rules:

triton_diag_ssm_backward_deduction

which is consistent with the implementation here, assuming we are looking at a portion of vectors with `offsets` as the indices.

**Review by breakds — COMMENTED** [review]
Finished first round of review and LGTM. A few readability comments.
Learned a lot from reading the code here.

File: alf/networks/s5/utils.py (hunk 1 above)
Comment (emailweixu) [reply alf/networks/s5/utils.py:51]: You are right. The code is from the original S5 repo, which is from https://github.com/srush/annotated-s4/blob/main/s4/s4.py

File: alf/networks/s5/utils.py (hunk 13 above)
Comment (emailweixu) [reply alf/networks/s5/utils.py]: Not only the line after 225, all the load/store of the imaginary part of a gradient are related to the comment.

File: alf/networks/s5/utils.py (hunk 6 above)
Comment (emailweixu) [reply alf/networks/s5/utils.py]: It is chosen from a few different values: 64,128,256

File: alf/networks/s5/utils.py (hunk 7 above)
Comment (emailweixu) [reply alf/networks/s5/utils.py]: done

File: alf/networks/s5/utils.py (hunk 12 above)
Comment (emailweixu) [reply alf/networks/s5/utils.py]: comment added.

File: alf/networks/s5/utils.py (hunk 11 above)
Comment (emailweixu) [reply alf/networks/s5/utils.py]: fixed

File: alf/networks/s5/s5.py (hunk 5 above)
Comment (emailweixu) [reply alf/networks/s5/s5.py:58]: Sometimes, it's more efficient to treat them as real. For example, if we only want real the part of `A * x` where A and x are complex.

File: alf/networks/s5/s5.py (hunk 10 above)
Comment (emailweixu) [reply alf/networks/s5/s5.py:119]: added

File: alf/networks/s5/s5.py (hunk 2 above)
Comment (emailweixu) [reply alf/networks/s5/s5.py:140]: added

File: alf/networks/s5/s5.py (hunk 9 above)
Comment (emailweixu) [reply alf/networks/s5/s5.py]: updated

File: alf/networks/s5/s5.py (hunk 3 above)
Comment (emailweixu) [reply alf/networks/s5/s5.py:167]: added

File: alf/networks/s5/s5.py (hunk 4 above)
Comment (emailweixu) [reply alf/networks/s5/s5.py:337]: added.

File: alf/networks/s5/utils.py [hunk 15]
```diff
@@ -83,248 +81,263 @@ def make_DPLR_HiPPO(N):
...
+            return grad_s, grad_x, grad_lambda.sum(dim=0)
+
+    diag_ssm_forward_triton = _ssm_forward.apply
+
+except ImportError:
+    from alf.utils.common import warning_once
+
+    def diag_ssm_forward_triton(s, x, Lambda):
+        warning_once("Triton is not installed. Using slow diag_ssm_forward.")
+        return diag_ssm_forward_slow(s, x, Lambda)
```
Comment (le-horizon) [inline alf/networks/s5/utils.py:340]: This has the disadvantage of falsely pass unit test when slow and triton paths actually differ.
Maybe try something like this:
https://github.com/HorizonRobotics/alf/commit/2ac807c39a1ca88bf36a51386b9630619966d1fe

Comment (emailweixu) [reply alf/networks/s5/utils.py:340]: If triton is installed, it is always tested. If triton is not installed, then there is no way to test.

Comment (le-horizon) [reply alf/networks/s5/utils.py:340]: If say our CI docker image accidentally removes triton, then the unit tests will silently pass, instead of raise an error.

Comment (emailweixu) [reply alf/networks/s5/utils.py:340]: Our CI docker only runs in CPU.

**Review by le-horizon — DISMISSED** [review]
LGTM.  Thanks for the changes!

**Review by breakds — APPROVED** [review]
